executors.md

```diff
@@ -0,0 +1,179 @@
+# Summary
+[summary]: #summary
+
+This RFC proposes a design for `futures-executors`, including both executor
+traits and built-in executors. In addition, it sets up a core expectation for all
+tasks that they are able to spawn additional tasks, while giving fine-grained
+control over what executor that spawning is routed to.
+
+NOTE: this RFC assumes that [RFC #2] is accepted.
+
+[RFC #2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+# Motivation
+[motivation]: #motivation
+
+This is a follow-up to the [Tokio Reform], [Futures 0.2], and [Task Context]
+RFCs, harmonizing their designs.
+
+[Tokio Reform]: https://github.com/tokio-rs/tokio-rfcs/pull/3
+[Futures 0.2]: https://github.com/rust-lang-nursery/futures-rfcs/pull/1
+[Task Context]: https://github.com/rust-lang-nursery/futures-rfcs/pull/2
+
+The design in this RFC has the following goals:
+
+- Add a core assumption that tasks are *always* able to spawn additional tasks,
+  avoiding the need to reflect this at the API level. (Note: this assumption is
+  only made in contexts that also assume an allocator.)
+
+- Provide fine-grained control over which executor is used to fulfill that assumption.
+
+- Provide reasonable built-in executors for both single-threaded and
+  multithreaded execution.
+
+It brings the futures library into line with the design principles of the [Tokio
+Reform], but also makes some improvements to the `current_thread` design.
+
+# Proposed design
+
+## The core executor abstraction: `Spawn`
+
+First, in `futures-core` we provide a basic definition of a general-purpose
+executor:
+
+```rust
+trait Spawn {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>);
```

Comment:
Is there a reason this isn't `&mut self`? I've been finding that using `&self` ended up hitting problems in a number of cases. Also, since this is stored on `Context`, you have mutable access anyway.

`Spawn` instances that do require `&self` can always impl `Spawn for &Self`, like `TcpStream` and co. w/ `Read`.